### PR TITLE
feat: atom onMount

### LINKED ISSUE #211: [RFC] atom effects
Continuing from discussions in the discord channel.
Recoil proposes [Atom Effects](https://recoiljs.org/docs/guides/atom-effects/).
We don't need to provide the same API, and probably it's impossible as there are some conceptual differences between jotai and recoil. (It is interesting to me that things get to be seen differently in jotai from recoil, because originally jotai is modeled after recoil.)

However, there are good use cases. It seems like there are three possibilities for each use case.
1. can be implemented as derived atoms
2. can't be implemented as derived atoms, but something can be added in jotai core
3. can never be implemented in jotai

Collecting use cases might be important. Please share your ideas.

For 1, we could create new functions in jotai/utils: `withEffect` and `atomWithEffect`.

Any thoughts? volunteers?

### LINKED ISSUE #234: Good pattern for updating atoms based on subscription
Lets say I have an API that I can subscribe to, and any time a value changes (can happen once a frame at 60FPS) I would want to update an atom to always reflect that

Currently I can just create a useEffect to do this at the App.tsx level, but wondering if there's an atoms-only way to do this.

some pseudocode:
```
useEffect(() => {
  async function load() {
    setAtom(client.get(someId))
  }
  client.subscribe(someId, (msg) => setAtom(msg));
  () => {
    client.unsubscribe(someId);
  }
}, [])
```

### PR DESCRIPTION
**Pull request by dai-shi**
close #211 
close #234 
related #263 
related #248 

- [x] first hypothetical implementation
- [x] second implementation `onMount`
- [x] third implementation `onMount(setAtom)`
- [x] add tests
- [x] fix implementation
  - mutual dependents (maybe not for this pr)
  - [x] `[Bug] a dependency is not mounted`
- [x] refactor code
- [x] docs

### CODE REVIEW

File: src/core/vanilla.ts [hunk 1]
```diff
@@ -336,9 +336,7 @@ export const delAtom = (
   if (mounted) {
     const [dependents] = mounted
     dependents.delete(useId)
-    if (!dependents.size) {
-      unmountAtom(state, deletingAtom)
-    }
```
Comment (dai-shi) [inline src/core/vanilla.ts]: @Aslemammad The intuition behind this is when this deletingAtom has other dependents, we shouldn't unmount it. Could you create some other test cases (I can think at least two cases (1) two components use an atom directly; (2) one component use it, another component use it indirectly), and see this change fails.

File: tests/onmount.test.tsx [hunk 2]
```diff
@@ -144,7 +144,7 @@ it('mount/unmount test', async () => {
   expect(onUnMountFn).toBeCalledTimes(1)
 })
 
-it('one derived atom, one onMount for the derived one, and one for the regular atom + onUnMount', async () => {
+it('derive chain', async () => {
```
Comment (dai-shi) [inline tests/onmount.test.tsx:147]: oh, we can leave the descriptive version. my list in the comment is just dumb.

Comment (Aslemammad) [reply tests/onmount.test.tsx:147]: Aha, Ok

**Comment by dai-shi** [comment]
Made an example in which an atom is sync with redux store.
https://codesandbox.io/s/react-typescript-forked-pftk3?file=/src/App.tsx

We can even sync the whole store and use `useSelector` from `'jotai/utils'`. 😲 
(and create dispatchAtom for `useDispatch`. In essence, creating `react-redux` with `jotai`. )

File: .size-snapshot.json [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {
   "index.js": {
-    "bundled": 17865,
-    "minified": 8067,
-    "gzipped": 2823,
+    "bundled": 19712,
+    "minified": 8864,
+    "gzipped": 3071,
```
Comment (dai-shi) [inline .size-snapshot.json:5]: I don't know why it increases so much.

Comment (Aslemammad) [reply .size-snapshot.json:5]: Sorry, my wrong, Let me fix it.

Comment (dai-shi) [reply .size-snapshot.json:5]: I don't think so, I run quite lately.

Comment (Aslemammad) [reply .size-snapshot.json:5]: Wow, that's weird, the same thing, I thought it's my old build problem.

**Review by dai-shi — COMMENTED** [review]
Should be good to go.

**Pull request merged**